Zero-initialize object factory structs

smowton opened the pull request:
Previously these were incrementally populated (assigning each field in turn); the zero init
was skipped because it is technically redundant. However, this prevents symex (and perhaps other
analyses) from propagating useful information, as the first write appears to be a partial update
on top of uninitialised data, and each subsequent write is based on its predecessor. Hence objects
produced by the factory end up represented as a stack of WITH operations, ultimately based on an
undefined symbol (e.g. dynamic_object1#0).

With this change symex becomes able to constant propagate the initial object, reducing equation
complexity, and objects that don't have any fields to nondet initialise can potentially be constant-
propagated throughout their lifetime.

tautschnig commented:
May I ask to take a look at #35, which introduces a variant of `zero_initializer` that produces a nondet element? It would also solve this problem not only for Java...

smowton commented:
Looked at #35 -- the nondet-initializer looks useful; this PR would then use it where it now uses `zero_initializer`. It wouldn't make `java_object_factoryt` go away, as `nondet_initializer` doesn't do anything special for pointers, but it would mean the object factory would *only* be concerned with pointer members, which is an improvement.

The symex tweaks to skip nondet initialization on certain declarations I had trouble understanding however.

smowton committed:
Zero-initialize object factory structs

Previously these were incrementally populated (assigning each field in turn); the zero init
was skipped because it is technically redundant. However, this prevents symex (and perhaps other
analyses) from propagating useful information, as the first write appears to be a partial update
on top of uninitialised data, and each subsequent write is based on its predecessor. Hence objects
produced by the factory end up represented as a stack of WITH operations, ultimately based on an
undefined symbol (e.g. dynamic_object1#0).

With this change symex becomes able to constant propagate the initial object, reducing equation
complexity, and objects that don't have any fields to nondet initialise can potentially be constant-
propagated throughout their lifetime.

smowton commented:
Rebased this and amended to obey `skip_classid` (used in Deeptest, and shortly in jbmc) and `update_in_place`, both of which are reasons not to make a blanket whole-object write.

This should be tested with Deeptest once it can work with mainline cbmc again.
